src/lib/util/artifact_download.js

```diff
@@ -29,78 +31,58 @@ module.exports = async function(queue, stream, taskId, artifactPath, destination
     taskId,
     artifactPath
   );
-  let attempts = 0;
+
+  // As we change the semantics of these parameters, we add
+  // some preventive checks in case we didn't handle all the cases
+  // in production.
+  if (delayFactor > 10) {
+    delayFactor = RETRY_CONFIG.delayFactor;
+  }
+  if (typeof randomizationFactor !== 'boolean') {
+    randomizationFactor = RETRY_CONFIG.randomizationFactor;
+  }
 
   stream.write(
     fmtLog(`Downloading artifact "${artifactPath}" from task ID: ${taskId}.`)
   );
-  while (attempts++ < maxAttempts) {
-    let hash = crypto.createHash('sha256');
-    let destinationStream = fs.createWriteStream(destination);
-    try {
-      let expectedSize = 0;
-      let receivedSize;
-      let req = request.get(artifactUrl);
-      req.on('response', (res) => {
-        expectedSize = parseInt(res.headers['content-length']);
-        receivedSize = 0;
-      });
-      req.on('data', (chunk) => {
-        receivedSize += chunk.length;
-        hash.update(chunk);
-      });
 
-      let intervalId = setInterval(() => {
-        if (receivedSize) {
-          stream.write(fmtLog(
-            `Download Progress: ${((receivedSize / expectedSize) * 100).toFixed(2)}%`
-          ));
-        }
-      }, 5000);
-
-      try {
-        await pipe(req, destinationStream);
-      } finally {
-        clearInterval(intervalId);
-      }
+  let bar = new ProgressBar({
+    schema: 'Download progress :percent [:bar]',
+    filled: '=',
+    blank: ' ',
+  });
 
-      if (req.response.statusCode !== 200) {
-        let error = new Error(req.response.statusMessage);
-        error.statusCode = req.response.statusCode;
-        throw error;
-      }
+  return await promiseRetry((retry, attempt) => {
+    return new Promise(async(accept, reject) => {
+      let hash = crypto.createHash('sha256');
+      let destinationStream = fs.createWriteStream(destination);
 
-      if (receivedSize !== expectedSize) {
-        throw new Error(`Expected size is '${expectedSize}' but received '${receivedSize}'`);
-      }
+      let sink = got.stream(artifactUrl)
+        .on('data', chunk => hash.update(chunk))
+        .on('error', reject)
+        .on('downloadProgress', progress => bar.update(progress.percent));
+      await pipe(sink, destinationStream);
```

Comment:
My understanding of streams is terrible, but .. don't you need to wait for this stream to complete?  I don't think `pipe` actually returns anything, so awaiting it doesn't do so.  But maybe I'm wrong?

Reply:
This uses the [promisepipe](https://github.com/epeli/node-promisepipe) package, the `await` keyword handles it.

Reply:
Ah, indeed, I am wrong :)